File: source/lang/lexer.py
"""Lexer for the visual DSL."""

import re
from typing import Iterator
from .types import Token, TokenType


class Lexer:
    """Tokenize DSL source into a stream of tokens."""

    PATTERNS = [
        # Comments (skip)
        (r'//[^\n]*', None),
        # Variables
        (r'\$[a-zA-Z_][a-zA-Z0-9_]*', TokenType.VAR),
        # Colors
        (r'#[0-9a-fA-F]{3,8}\b', TokenType.COLOR),
        # Pairs: 100,200 or 100x200
        (r'-?\d+\.?\d*[,x]-?\d+\.?\d*', TokenType.PAIR),
        # Strings
        (r'"[^"]*"', TokenType.STRING),
        (r"'[^']*'", TokenType.STRING),
        # Numbers
        (r'-?\d+\.?\d*', TokenType.NUMBER),
        # Operators
        (r'->', TokenType.ARROW),
        (r':', TokenType.COLON),
        (r'=', TokenType.EQUALS),
        # Identifiers
        (r'[a-zA-Z_][a-zA-Z0-9_-]*', TokenType.IDENT),
    ]

    def __init__(self, source: str):
        self.source = source
        self.lines = source.split('\n')
        self.indent_stack = [0]

    def tokenize(self) -> Iterator[Token]:
        """Generate tokens from source."""
        for lineno, line in enumerate(self.lines):
            stripped = line.lstrip()
            
            # Skip empty and comment-only lines
            if not stripped or stripped.startswith('//'):
                continue

            indent = len(line) - len(stripped)
            yield from self._handle_indent(indent, lineno)
            yield from self._tokenize_line(stripped, lineno)
            yield Token(TokenType.NEWLINE, '\n', lineno, len(line))

        # Close remaining indents
        while len(self.indent_stack) > 1:
            self.indent_stack.pop()
            yield Token(TokenType.DEDENT, '', len(self.lines) - 1, 0)

        yield Token(TokenType.EOF, '', len(self.lines) - 1, 0)

    def _handle_indent(self, indent: int, line: int) -> Iterator[Token]:
        """Handle indentation changes."""
        if indent > self.indent_stack[-1]:
            self.indent_stack.append(indent)
            yield Token(TokenType.INDENT, '', line, 0)
        else:
            while indent < self.indent_stack[-1]:
                self.indent_stack.pop()
                yield Token(TokenType.DEDENT, '', line, 0)

    def _tokenize_line(self, line: str, lineno: int) -> Iterator[Token]:
        """Tokenize a single line."""
        pos = 0
        while pos < len(line):
            # Skip whitespace
            if line[pos].isspace():
                pos += 1
                continue

            matched = False
            for pattern, ttype in self.PATTERNS:
                if m := re.match(pattern, line[pos:]):
                    if ttype is not None:  # Skip comments
                        val = self._parse_value(m.group(), ttype)
                        yield Token(ttype, val, lineno, pos)
                    pos += len(m.group())
                    matched = True
                    break
            
            if not matched:
                pos += 1  # Skip unknown char

    def _parse_value(self, raw: str, ttype: TokenType):
        """Parse token value based on type."""
        match ttype:
            case TokenType.NUMBER:
                return float(raw) if '.' in raw else int(raw)
            case TokenType.STRING:
                return raw[1:-1]  # Strip quotes
            case TokenType.PAIR:
                sep = 'x' if 'x' in raw else ','
                a, b = raw.split(sep)
                return (float(a) if '.' in a else int(a),
                        float(b) if '.' in b else int(b))
            case _:
                return raw

File: distribution/npm/src/core/lexer.ts
import type { Token, TokenType } from './types';

type PatternEntry = [RegExp, TokenType | null];

const PATTERNS: PatternEntry[] = [
  [/^\/\/[^\n]*/, null], // Comments (skip)
  [/^\$[a-zA-Z_][a-zA-Z0-9_]*/, 'VAR'],
  [/^#[0-9a-fA-F]{3,8}\b/, 'COLOR'],
  [/^-?\d+\.?\d*[,x]-?\d+\.?\d*/, 'PAIR'],
  [/^"[^"]*"/, 'STRING'],
  [/^'[^']*'/, 'STRING'],
  [/^-?\d+\.?\d*/, 'NUMBER'],
  [/^\[/, 'LBRACKET'],
  [/^\]/, 'RBRACKET'],
  [/^->/, 'ARROW'],
  [/^:/, 'COLON'],
  [/^=/, 'EQUALS'],
  [/^[a-zA-Z_][a-zA-Z0-9_-]*/, 'IDENT'],
];

export class Lexer {
  private lines: string[];
  private indentStack: number[] = [0];

  constructor(source: string) {
    this.lines = source.split('\n');
  }

  *tokenize(): Generator<Token> {
    for (let lineno = 0; lineno < this.lines.length; lineno++) {
      const line = this.lines[lineno];
      const stripped = line.trimStart();

      if (!stripped || stripped.startsWith('//')) continue;

      const indent = line.length - stripped.length;
      yield* this.handleIndent(indent, lineno);
      yield* this.tokenizeLine(stripped, lineno);
      yield { type: 'NEWLINE', value: '\n', line: lineno, col: line.length };
    }

    while (this.indentStack.length > 1) {
      this.indentStack.pop();
      yield { type: 'DEDENT', value: '', line: this.lines.length - 1, col: 0 };
    }

    yield { type: 'EOF', value: '', line: this.lines.length - 1, col: 0 };
  }

  private *handleIndent(indent: number, line: number): Generator<Token> {
    const current = this.indentStack[this.indentStack.length - 1];
    if (indent > current) {
      this.indentStack.push(indent);
      yield { type: 'INDENT', value: '', line, col: 0 };
    } else {
      while (indent < this.indentStack[this.indentStack.length - 1]) {
        this.indentStack.pop();
        yield { type: 'DEDENT', value: '', line, col: 0 };
      }
    }
  }

  private *tokenizeLine(line: string, lineno: number): Generator<Token> {
    let pos = 0;
    while (pos < line.length) {
      if (/\s/.test(line[pos])) { pos++; continue; }

      let matched = false;
      for (const [pattern, ttype] of PATTERNS) {
        const match = line.slice(pos).match(pattern);
        if (match) {
          if (ttype !== null) {
            yield { type: ttype, value: this.parseValue(match[0], ttype), line: lineno, col: pos };
          }
          pos += match[0].length;
          matched = true;
          break;
        }
      }
      if (!matched) pos++;
    }
  }

  private parseValue(raw: string, ttype: TokenType): string | number | [number, number] {
    switch (ttype) {
      case 'NUMBER':
        return raw.includes('.') ? parseFloat(raw) : parseInt(raw, 10);
      case 'STRING':
        return raw.slice(1, -1);
      case 'PAIR': {
        const sep = raw.includes('x') ? 'x' : ',';
        const [a, b] = raw.split(sep);
        return [parseFloat(a), parseFloat(b)];
      }
      default:
        return raw;
    }
  }
}

